src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,100 +141,125 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
         }
     }
 
+    protected abstract Task<(RemoteProjectFile? projectFile, bool hasAllInformation, BuildHostProcessKind preferred, BuildHostProcessKind actual)> TryLoadProjectAsync(
+        BuildHostProcessManager buildHostProcessManager, string projectPath, CancellationToken cancellationToken);
+
     /// <returns>True if the project needs a NuGet restore, false otherwise.</returns>
     private async Task<bool> LoadOrReloadProjectAsync(ProjectToLoad projectToLoad, ToastErrorReporter toastErrorReporter, BuildHostProcessManager buildHostProcessManager, CancellationToken cancellationToken)
     {
         BuildHostProcessKind? preferredBuildHostKindThatWeDidNotGet = null;
         var projectPath = projectToLoad.Path;
+        Contract.ThrowIfFalse(PathUtilities.IsAbsolute(projectPath));
 
         try
         {
-            var preferredBuildHostKind = GetKindForProject(projectPath);
-            var (buildHost, actualBuildHostKind) = await buildHostProcessManager.GetBuildHostWithFallbackAsync(preferredBuildHostKind, projectPath, cancellationToken);
-            if (preferredBuildHostKind != actualBuildHostKind)
-                preferredBuildHostKindThatWeDidNotGet = preferredBuildHostKind;
-
-            if (!_projectFileExtensionRegistry.TryGetLanguageNameFromProjectPath(projectPath, DiagnosticReportingMode.Ignore, out var languageName))
-                return false;
-
-            var loadedFile = await buildHost.LoadProjectFileAsync(projectPath, languageName, cancellationToken);
-            var diagnosticLogItems = await loadedFile.GetDiagnosticLogItemsAsync(cancellationToken);
-            if (diagnosticLogItems.Any(item => item.Kind is DiagnosticLogItemKind.Error))
+            if (!_loadedProjects.TryGetValue(projectPath, out var loadedProjectSet) || loadedProjectSet.CancellationTokenSource.IsCancellationRequested)
             {
-                await LogDiagnosticsAsync(diagnosticLogItems);
-                // We have total failures in evaluation, no point in continuing.
+                // project was already unloaded or in process of unloading.
                 return false;
             }
 
-            var loadedProjectInfos = await loadedFile.GetProjectFileInfosAsync(cancellationToken);
-
-            // The out-of-proc build host supports more languages than we may actually have Workspace binaries for, so ensure we can actually process that
-            // language in-process.
-            var projectLanguage = loadedProjectInfos.FirstOrDefault()?.Language;
-            if (projectLanguage != null && ProjectFactory.Workspace.Services.GetLanguageService<ICommandLineParserService>(projectLanguage) == null)
+            var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(loadedProjectSet.CancellationTokenSource.Token, cancellationToken);
```

Comment:
Most of the unloading code is making sense to me at this point, except:

It looks like this is not actually used here. I believe the param `cancellationToken` represents whether the queue itself has been disposed, and `loadedProjectSet.CancellationTokenSource` represents whether the set is being disposed.

I do not think we have any existing path which actually disposes the whole queue. I think only the LS.exe shutting down would do that. In which case I don't think we need to go through and dispose the projects.

So, possibly, we should just delete this, and when we try to catch OperationCancelledException we filter to make sure it is for `loadedProjectSet.CancellationTokenSource` instead.

Curious what you think @jasonmalinowski.

Reply:
Moved to https://github.com/dotnet/roslyn/pull/78488#discussion_r2078185546